File: bot/api/lifi.py
"""LI.FI API client."""
from __future__ import annotations

from typing import Any

import httpx

from config import LIFI_BASE_URL, LIFI_API_KEY


async def fetch_quote(
    from_chain: str,
    to_chain: str,
    from_token: str,
    to_token: str,
    from_amount: str,
    from_address: str,
    to_address: str | None = None,
) -> dict[str, Any]:
    """Fetch a swap/bridge quote from LI.FI."""
    params: dict[str, str] = {
        "fromChain": from_chain,
        "toChain": to_chain,
        "fromToken": from_token,
        "toToken": to_token,
        "fromAmount": from_amount,
        "fromAddress": from_address,
    }
    if to_address:
        params["toAddress"] = to_address

    headers: dict[str, str] = {}
    if LIFI_API_KEY:
        headers["x-lifi-api-key"] = LIFI_API_KEY

    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.get(
            f"{LIFI_BASE_URL}/v1/quote",
            params=params,
            headers=headers,
        )
        resp.raise_for_status()
        return resp.json()

File: web/src/api/lifi.ts
import type { LifiQuote } from '../types';

const LIFI_BASE_URL = import.meta.env.VITE_LIFI_BASE_URL || 'https://li.quest';
const LIFI_API_KEY = import.meta.env.VITE_LIFI_API_KEY || '';

export interface QuoteParams {
  fromChain: string;
  toChain: string;
  fromToken: string;
  toToken: string;
  fromAmount: string;
  fromAddress: string;
  toAddress?: string;
}

export async function fetchLifiQuote(params: QuoteParams): Promise<LifiQuote> {
  const searchParams = new URLSearchParams({
    fromChain: params.fromChain,
    toChain: params.toChain,
    fromToken: params.fromToken,
    toToken: params.toToken,
    fromAmount: params.fromAmount,
    fromAddress: params.fromAddress,
  });
  if (params.toAddress) {
    searchParams.set('toAddress', params.toAddress);
  }

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };
  if (LIFI_API_KEY) {
    headers['x-lifi-api-key'] = LIFI_API_KEY;
  }

  const res = await fetch(`${LIFI_BASE_URL}/v1/quote?${searchParams.toString()}`, {
    headers,
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`LI.FI API error ${res.status}: ${text}`);
  }

  return res.json() as Promise<LifiQuote>;
}
